File: benchmarks/utf8/main.cpp
// utf8 — dựng một chuỗi nhiều ngôn ngữ rồi duyệt qua từng ký tự Unicode.
//
// Đây là chỗ bốn ngôn ngữ hiểu "một ký tự" theo bốn kiểu khác hẳn nhau: Go là chuỗi byte
// duyệt ra rune, Rust là UTF-8 có kiểm biên, JS là UTF-16 nên ký tự ngoài BMP chiếm hai
// đơn vị, còn C++ không có khái niệm Unicode nào cả — phải tự giải mã bằng tay.
//
// Chuỗi sinh từ cùng một LCG nên bốn ngôn ngữ nhận đúng cùng dãy điểm mã. Trọng số dùng
// thứ tự của ĐIỂM MÃ, không phải vị trí byte — nếu không thì bốn ngôn ngữ đánh số khác nhau.
//
// utf8 — build a multilingual string, then walk it character by character.
//
// This is where the four languages disagree most about what "a character" is: Go is bytes
// walked as runes, Rust is UTF-8 with boundary checks, JS is UTF-16 so anything outside
// the BMP takes two units, and C++ has no notion of Unicode at all — it must decode by hand.
//
// The string comes from the same LCG, so all four see the same code points. The weight is
// the CODE POINT index, never a byte offset, which the four would number differently.
#include "common.hpp"
#include <string>
#include <vector>

int main(int argc, char** argv) {
  int n = (int)lb::param(argc, argv, "n", 400000);

  // Bốn dải: ASCII 1 byte, Latin 2 byte, CJK 3 byte, ký hiệu 4 byte.
  // Four ranges: ASCII at 1 byte, Latin at 2, CJK at 3, symbols at 4.
  std::vector<uint32_t> cps((size_t)n);
  lb::Lcg rng(2024u);
  for (int i = 0; i < n; ++i) {
    uint32_t r = rng.next();
    switch (r & 3u) {
      case 0: cps[(size_t)i] = 0x20u + (r >> 8) % 95u; break;
      case 1: cps[(size_t)i] = 0xC0u + (r >> 8) % 64u; break;
      case 2: cps[(size_t)i] = 0x4E00u + (r >> 8) % 0x5000u; break;
      default: cps[(size_t)i] = 0x1F300u + (r >> 8) % 0x300u; break;
    }
  }

  std::string s;
  s.reserve((size_t)n * 3);
  for (uint32_t cp : cps) {
    if (cp < 0x80) {
      s.push_back((char)cp);
    } else if (cp < 0x800) {
      s.push_back((char)(0xC0 | (cp >> 6)));
      s.push_back((char)(0x80 | (cp & 0x3F)));
    } else if (cp < 0x10000) {
      s.push_back((char)(0xE0 | (cp >> 12)));
      s.push_back((char)(0x80 | ((cp >> 6) & 0x3F)));
      s.push_back((char)(0x80 | (cp & 0x3F)));
    } else {
      s.push_back((char)(0xF0 | (cp >> 18)));
      s.push_back((char)(0x80 | ((cp >> 12) & 0x3F)));
      s.push_back((char)(0x80 | ((cp >> 6) & 0x3F)));
      s.push_back((char)(0x80 | (cp & 0x3F)));
    }
  }

  lb::Timer t;

  uint32_t sum = 0, wide = 0, idx = 0;
  size_t i = 0;
  while (i < s.size()) {
    uint8_t b0 = (uint8_t)s[i];
    uint32_t cp;
    if (b0 < 0x80) { cp = b0; i += 1; }
    else if ((b0 & 0xE0) == 0xC0) { cp = ((uint32_t)(b0 & 0x1F) << 6) | ((uint8_t)s[i+1] & 0x3F); i += 2; }
    else if ((b0 & 0xF0) == 0xE0) {
      cp = ((uint32_t)(b0 & 0x0F) << 12) | (((uint8_t)s[i+1] & 0x3F) << 6) | ((uint8_t)s[i+2] & 0x3F);
      i += 3;
    } else {
      cp = ((uint32_t)(b0 & 0x07) << 18) | (((uint8_t)s[i+1] & 0x3F) << 12)
         | (((uint8_t)s[i+2] & 0x3F) << 6) | ((uint8_t)s[i+3] & 0x3F);
      i += 4;
    }
    idx += 1;
    sum += idx * cp;
    if (cp > 0x7F) wide += 1;
  }

  double ms = t.ms();
  lb::Checksum c;
  c.add(sum);
  c.add(wide);
  c.add(idx);
  c.add_u64((uint64_t)s.size());
  lb::report(ms, lb::hex8(c.value()));
}

File: benchmarks/utf8/main.pas
// utf8 — dựng một chuỗi nhiều ngôn ngữ rồi duyệt qua từng ký tự Unicode.
//
// Pascal giống C++ ở chỗ AnsiString chỉ là dãy byte, không có khái niệm Unicode nào — nên
// phần giải mã UTF-8 phải tự viết. Trọng số dùng thứ tự của ĐIỂM MÃ, không phải vị trí byte.
//
// utf8 — build a multilingual string, then walk it character by character.
//
// Pascal is like C++ here: an AnsiString is just bytes with no notion of Unicode, so the
// UTF-8 decoding is written by hand. The weight is the CODE POINT index, never a byte offset.
program Utf8Walk;
{$mode objfpc}{$H+}{$Q-}{$R-}
uses Common;

var
  N, I, Pos, Len: LongInt;
  Cps: array of LongWord;
  Rng: TLcg;
  R, Cp, Sum, Wide, Idx: LongWord;
  S: AnsiString;
  B0: Byte;
  T: TTimer;
  Ms: Double;
  C: TChecksum;
begin
  N := Param('n', 6000000);

  // Bốn dải: ASCII 1 byte, Latin 2 byte, CJK 3 byte, ký hiệu 4 byte.
  // Four ranges: ASCII at 1 byte, Latin at 2, CJK at 3, symbols at 4.
  SetLength(Cps, N);
  Rng.Init(2024);
  for I := 0 to N - 1 do
  begin
    R := Rng.Next;
    case R and 3 of
      0: Cps[I] := $20 + (R shr 8) mod 95;
      1: Cps[I] := $C0 + (R shr 8) mod 64;
      2: Cps[I] := $4E00 + (R shr 8) mod $5000;
    else
      Cps[I] := $1F300 + (R shr 8) mod $300;
    end;
  end;

  SetLength(S, N * 4);
  Pos := 1;
  for I := 0 to N - 1 do
  begin
    Cp := Cps[I];
    if Cp < $80 then
    begin
      S[Pos] := Chr(Cp); Inc(Pos);
    end
    else if Cp < $800 then
    begin
      S[Pos] := Chr($C0 or (Cp shr 6)); Inc(Pos);
      S[Pos] := Chr($80 or (Cp and $3F)); Inc(Pos);
    end
    else if Cp < $10000 then
    begin
      S[Pos] := Chr($E0 or (Cp shr 12)); Inc(Pos);
      S[Pos] := Chr($80 or ((Cp shr 6) and $3F)); Inc(Pos);
      S[Pos] := Chr($80 or (Cp and $3F)); Inc(Pos);
    end
    else
    begin
      S[Pos] := Chr($F0 or (Cp shr 18)); Inc(Pos);
      S[Pos] := Chr($80 or ((Cp shr 12) and $3F)); Inc(Pos);
      S[Pos] := Chr($80 or ((Cp shr 6) and $3F)); Inc(Pos);
      S[Pos] := Chr($80 or (Cp and $3F)); Inc(Pos);
    end;
  end;
  SetLength(S, Pos - 1);
  Len := Length(S);

  T.Start;

  Sum := 0; Wide := 0; Idx := 0;
  I := 1;
  while I <= Len do
  begin
    B0 := Byte(S[I]);
    if B0 < $80 then
    begin
      Cp := B0;
      Inc(I);
    end
    else if (B0 and $E0) = $C0 then
    begin
      Cp := (LongWord(B0 and $1F) shl 6) or (Byte(S[I + 1]) and $3F);
      I := I + 2;
    end
    else if (B0 and $F0) = $E0 then
    begin
      Cp := (LongWord(B0 and $0F) shl 12) or (LongWord(Byte(S[I + 1]) and $3F) shl 6)
         or (Byte(S[I + 2]) and $3F);
      I := I + 3;
    end
    else
    begin
      Cp := (LongWord(B0 and $07) shl 18) or (LongWord(Byte(S[I + 1]) and $3F) shl 12)
         or (LongWord(Byte(S[I + 2]) and $3F) shl 6) or (Byte(S[I + 3]) and $3F);
      I := I + 4;
    end;
    Inc(Idx);
    Sum := Sum + Idx * Cp;
    if Cp > $7F then Inc(Wide);
  end;

  Ms := T.Ms;
  C.Init;
  C.Add(Sum);
  C.Add(Wide);
  C.Add(Idx);
  C.AddU64(QWord(Len));
  Report(Ms, C.Hex);
end.
